vignettes/fims-path-maturity.Rmd

```diff
@@ -0,0 +1,614 @@
+//file: inst/include/common/information.hpp
+/*
+ uint32_t is an unsigned integer (always positive)
+ The first component of the map is a uint32_t which 
+ will be used to hold the ID.
+ The second component of the map is the shared pointer, 
+ maturity_models, that points to fims::MaturityBase
+ */
+std::map<uint32_t, std::shared_ptr<fims::MaturityBase<T> > >
+      maturity_models; /*!<hash map to link each object to its shared location
+                          in memory*/
+
+//Next we set up the iterator, which will be used to loop 
+//over all defined maturity modules
+typedef typename std::map<uint32_t, 
+                          std::shared_ptr<fims::MaturityBase<T> > >::iterator
+      maturity_models_iterator;
+```
+
+```{r image-path-maturity-5, echo = FALSE, message=FALSE, out.width = '100%'}
+knitr::include_graphics("figures/fims-path-maturity-5.png")
+```
+
+Next, let's revisit the line of code that was written in the Rcpp `LogisticMaturityInterface` class in [rcpp_maturity.hpp](https://github.com/NOAA-FIMS/FIMS/blob/main/inst/include/interface/rcpp/rcpp_objects/rcpp_maturity.hpp) under [inst/include/interface/rcpp/rcpp_objects](https://github.com/NOAA-FIMS/FIMS/tree/main/inst/include/interface/rcpp/rcpp_objects)
+
+Here, we are setting the maturity_models pointer in `Information` to equal the `lm0` pointer to the `LogisticMaturity` module
+```{Rcpp,  class.source = "rcppchunk", eval = FALSE}
+//file: rcpp_maturity.hpp
+
+/*
+ the maturity_models pointer from Information that matches the id of the 
+ fims::LogisticMaturity class is equal to the pointer to fims::LogisticMaturity 
+ */
+d0->maturity_models[lm0->id] = lm0;
+```
+
+```{r image-path-maturity-6, echo = FALSE, message=FALSE, out.width = '100%'}
+knitr::include_graphics("figures/fims-path-maturity-6.png")
+```
+
+Now we need to pass this pointer to the maturity pointer in population so that `population->maturity` points to the `LogisticMaturity` module instead of `MaturityBase`. First we set up a map in `Information` that points to `Population`. 
+```{Rcpp,  class.source = "infochunk", eval = FALSE}
+//file: inst/include/common/information.hpp
+std::map<uint32_t, std::shared_ptr<fims::Population<T> > >
+      populations; /*!<hash map to link each object to its shared location in
+                      memory*/
+  typedef typename std::map<uint32_t,
+                            std::shared_ptr<fims::Population<T> > >::iterator
+      population_iterator;
+  /**< iterator for population objects>*/
+```
+
+```{r image-path-maturity-7, echo = FALSE, message=FALSE, out.width = '100%'}
+knitr::include_graphics("figures/fims-path-maturity-7.png")
+```
+
+We then loop through the populations and create a new shared pointer, `p`, to reference the individual population of interest.  
+
+```{Rcpp,  class.source = "infochunk", eval = FALSE}
+//file: inst/include/common/information.hpp
+ for (population_iterator it = this->populations.begin();
```

Comment:
fix: I think @Andrea-Havron-NOAA mentioned that we are no longer using `it` because it can be negative?

Reply:
We are no longer using `int` in for loops because `int` can be negative. [Issue #339](https://github.com/NOAA-FIMS/FIMS/issues/379) is about the use of iterators (`it`) , we may simplify maps with iterators to vectors that can be referenced by their index directly.